ext/liquid_c/variable_lookup.c

```diff
@@ -1,10 +1,14 @@
 #include "liquid.h"
 #include "context.h"
 
-static ID id_has_key, id_aref, id_fetch;
+static ID id_has_key, id_aref, id_fetch, id_to_liquid_value;
 
 VALUE variable_lookup_key(VALUE context, VALUE object, VALUE key, bool is_command)
 {
+    if (key && rb_respond_to(key, id_to_liquid_value)) {
+        key = rb_funcall(key, id_to_liquid_value, 0);
+    }
```

Comment:
I don't see the need to check if C is non-zero.  It should be usable with rb_respond_to no matter what ruby value it is.

Also, you can use `rb_check_funcall` to call a method if it exists, which should avoid having to have the method get looked up twice to call it.

```suggestion
    VALUE key_value = rb_check_funcall(key, id_to_liquid_value, 0);
    if (key_value != Qundef)
        key = key_value;
```

Reply:
I think the common case is when the key is a string.  We may want to leverage that fast path only trying to call the method if `rb_obj_class(key) != rb_cString`.

A more important optimization would be to avoid this check entirely if it is a constant key, since there is already special cases for that for the `OP_LOOKUP_CONST_KEY` and `OP_LOOKUP_COMMAND` VM instructions.